server/src/instant/grouped_queue.clj

```diff
@@ -0,0 +1,56 @@
+(ns instant.grouped-queue
+  (:import
+   (java.util.concurrent LinkedBlockingQueue TimeUnit)
+   (clojure.lang PersistentQueue)))
+
+(def empty-q PersistentQueue/EMPTY)
+
+(defn create [{:keys [group-fn]}]
+  {:group-fn group-fn
+   :group-key->subqueue (atom {})
+   :main-queue (LinkedBlockingQueue.)})
+
+(defn enqueue! [{:keys [group-fn main-queue group-key->subqueue]
+                 :as _grouped-q} item]
+  (let [group-key (group-fn item)]
+    (if (nil? group-key)
+      ;; This item is not to be grouped.
+      (.put main-queue [:item item])
```

Comment:
I am thinking: 

Perhaps we should have a separate `ungrouped-queue`.

The main `main-queue` would only keep a 'signal', and become more of a `dispatch-queue`. 

The reason I am thinking of doing this: 

Right now, it would be hard to determine a true `queue-size`. If we had items stored in a separate queue, than we could derive `queue-size` by adding up all the respective child queues.

wdyt @dwwoelfel ?

Reply:
Could you just keep a counter that you increment when you add an item to the queue and decrement when you start processing an item?

Reply:
Honestly good point; was trying to be 'functional' about it; but the counter feels right. Much less indirection, and only a few places where we have to deal with it.

Will update